File: algoritmos/orden/quicksort/orden.c
#include <stdio.h>
#include <stdlib.h>

void swap(int* a, int* b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}

int partition(int arr[], int low, int high) {
    int pivot = arr[high];
    int i = (low - 1);

    for (int j = low; j < high; j++) {
        if (arr[j] <= pivot) {
            i++;
            swap(&arr[i], &arr[j]);
        }
    }
    swap(&arr[i + 1], &arr[high]);
    return (i + 1);
}

void quickSort(int arr[], int low, int high) {
    if (low < high) {
        int pi = partition(arr, low, high);
        quickSort(arr, low, pi - 1);
        quickSort(arr, pi + 1, high);
    }
}

int readRandomBin(int **array_int, int *num){
    FILE *file = fopen("../random_list.bin", "rb");
    if (file == NULL) {
        return 1;
    }

    fseek(file, 0, SEEK_END);
    long file_size = ftell(file);
    fseek(file, 0, SEEK_SET);

    *num = file_size / sizeof(int);
    *array_int = (int *)malloc(*num * sizeof(int));

    fread(*array_int, sizeof(int), *num, file);

    fclose(file);
    return 0;
}

int createRandomFile(int *array_int, int num){
    FILE *txt_file = fopen("random_list_c.txt", "w");
    if (txt_file == NULL) {
        return 1;
    }

    for (int i = 0; i < num; i++) {
        fprintf(txt_file, "%d\n", array_int[i]);
    }

    fclose(txt_file);
    return 0;
}

int createSortedFile(int *array_int, int num){
    FILE *txt_file_sorted = fopen("sorted_list_c.txt", "w");
    if (txt_file_sorted == NULL) {
        return 1;
    }

    for (int i = 0; i < num; i++) {
        fprintf(txt_file_sorted, "%d\n", array_int[i]);
    }

    fclose(txt_file_sorted);
    return 0;
}
    
int main() {

    int num;
    int *array_int;
    // archivo original
    if (readRandomBin(&array_int, &num) != 0){
        perror("Error al leer");
        return 1;
    }

    /*
    // txt para comprobar
    if (createRandomFile(array_int, num) != 0){
        perror("Error al crear el archivo de texto");
        free(array_int);
        return 1;
    }
    */

    quickSort(array_int, 0, num - 1);

    // txt para comprobar
    if (createSortedFile(array_int, num) != 0) {
        perror("Error al crear el archivo de texto");
        free(array_int);
        return 1;
    }
    
    free(array_int);

    return 0;
}

File: algoritmos/orden/quicksort/orden.py

def swap(arr, a, b):
    arr[a], arr[b] = arr[b], arr[a]

def partition(arr, low, high):
    pivot = arr[high]
    i = low - 1

    for j in range(low, high):
        if arr[j] <= pivot:
            i += 1
            swap(arr, i, j)
    swap(arr, i + 1, high)
    return i + 1

def quickSort(arr, low, high):
    if low < high:
        pi = partition(arr, low, high)
        quickSort(arr, low, pi - 1)
        quickSort(arr, pi + 1, high)

def readRandomBin():
    try:
        with open("../random_list.bin", "rb") as file:
            array_bytes = file.read()
    except:
        print("Error al abrir")
        exit(1)

    num = len(array_bytes) // 4
    array_int = [int.from_bytes(array_bytes[i:i+4], byteorder='little') for i in range(0, len(array_bytes), 4)]

    return num, array_int

def createRandomFile(array_int):
    try:
        with open("random_list_py.txt", "w") as txt_file:
            for num_entero in array_int:
                txt_file.write(str(num_entero) + "\n")
    except:
        print("Error al crear el archivo de texto")
        exit(1)

def createSortedFile(array_int):
    try:
        with open("sorted_list_py.txt", "w") as txt_file_sorted:
            for num_entero in array_int:
                txt_file_sorted.write(str(num_entero) + "\n")
    except:
        print("Error al crear el archivo de texto")
        exit(1)

if __name__ == "__main__":

    # archivo original
    num, array_int = readRandomBin()        

    """
    # txt para comprobar
    createRandomFile(array_int)
    """
    
    quickSort(array_int, 0, num - 1)

    # txt para comprobar
    createSortedFile(array_int)
    